<!doctype html>
<html lang="en">
  {% include 'head.html' %}
  <style>
  ::-webkit-scrollbar {
      -webkit-appearance: none;
      width: 7px;
      height: 7px;
  }

  ::-webkit-scrollbar-thumb {
      border-radius: 4px;
      background-color: rgba(0,0,0,.5);
      -webkit-box-shadow: 0 0 1px rgba(255,255,255,.5);
  }
  </style>

  <body style="position: relative;">
    {% include 'top_menu_comparison.html' %}

    <div class="container-fluid variant_comparison">
      <div class="row">
    
        {% include 'sidebar_comparison.html' %}
        
        <div class="col-md-9 ml-sm-auto col-lg-10 px-md-4 mt-4">
          <h1 id="Summary">Summary</h1>
          <p>
            We calculated the numbers of tightly binding peptides (IC<sub>50</sub> affinity &leq; 50 nM)
            originating from the two selected variants for all entered alleles.
            The total set of tight binders was calculated as a union of peptide sets
            for each individual allele, i.e., a peptide with high binding affinity with 
            two entered alleles was counted one time.
          </p>
          <p>
            Aside from the binding affinity analysis, we also report worldwide frequencies of
            possible haplotypes, composed of entered alleles. Haplotype frequency data were
            imported from <a href="http://www.allelefrequencies.net" target="_blank">The Allele Frequency Net Database</a>.
            {% if hla_class == "HLA-II" %}
              Note that for HLA class II we present only HLA-DRB1/DQB1/DPB1 haplotype
              frequencies since the vast majority of haplotype frequency data include only these genes.
            {% endif %}
          </p>
          <table class="table table-bordered table-striped">
            <thead>
              <tr>
                <th class="align-top">HLA-{{ haplotype_frequency.index.name }}</th>
                {% for column in haplotype_frequency.columns %}
                  {% if column == "South and Central America" %}
                    {% set column = "S&C America" %}
                  {% endif %}
                  {% if column == "Weighted" %}
                    <th class="align-top">{{ column }} mean</th>
                  {% elif column == "Mean"%}
                    <th class="align-top">{{ column }}</td>
                  {% else %}
                    <td class="align-top">{{ column }}</td>
                  {% endif %}
                {% endfor %}
              </tr>
            </thead>
            <tbody>
              {% for haplotype, row in haplotype_frequency.iterrows(): %}  
                <tr>
                  <th>{{ haplotype }}</th>
                  {% for column in haplotype_frequency.columns %}
                    {% if column == "Weighted" %}
                      <th> {{ '%0.2f'|format(row[column]) }}</th>
                    {% elif column == "Mean"%}
                      <th> {{ '%0.2f'|format(row[column]) }}</th>
                    {% else %}
                      <td> {{ '%0.2f'|format(row[column]) }}</td>
                    {% endif %}
                  {% endfor %}
                </tr>
              {% endfor %}
            </tbody>
          </table>  

          {% for protein in proteins %}
            {% if protein == "All" %}
              <h1 id={{ protein }}>All proteins</h1>
            {% else %}
              <h1 id={{ protein }}>{{ protein[0]|upper }}{{ protein[1:] }} </h1>
            {% endif %}
            <a class="btn btn-link" style="float: right;" href="/comparison/{{ first_gisaid_id }}/{{ second_gisaid_id }}/download?hla_alleles={{hla_alleles}}&protein={{ protein }}" role="button" download>Export raw data to csv&nbsp;<i class="bi bi-download"></i></a>

            <table class="table table-bordered table-striped">
              <thead>
                <tr>
                  <th>Allele</th>
                  <th>{{ lineages.loc[lineages["GISAID Accession ID"] == first_gisaid_id]["Lineage"].iloc[0] }}</th>
                  <th>{{ lineages.loc[lineages["GISAID Accession ID"] == second_gisaid_id]["Lineage"].iloc[0] }}</th>
                </tr>
              </thead>
              <tbody>
              {% for _, row in hla_summary[protein].iterrows() %}
                {% if not (row['Allele'] == 'Summary') %}
                  <tr>
                    <td>{{ row['Allele'] }}</td>
                    <td>{{ row['First'] }}</td>
                    <td>{{ row['Second'] }}</td>
                  </tr>
                {% else %}
                  <tr>
                    <th>Total</th>
                    <th>{{ row['First'] }}</th>
                    <th>{{ row['Second'] }}</th>
                  </tr>
                {% endif %}
              {% endfor %}
              </tbody>
            </table>  
            {% set df = hla_summary[protein] %}
          {% endfor %}
        </div>
      </div>
    </div>

    {% include 'footer.html' %}

  </body>
</html>
